Test.hs

```diff
@@ -0,0 +1,61 @@
+import Tuura.Concurrency
+import Data.List
+import Data.Set (Set)
+import qualified Data.Set as Set
+import qualified Data.Map as Map
+import Tuura.Log
+
+readLog :: String -> IO (Log String)
+readLog strLog = do
+    return . canonicalise $ parse strLog
+
+toIntLog :: Log String -> (Log Int, Set Int, Int -> String)
+toIntLog log1 = (intLog, Set.map a2i alphabet, i2a)
+  where
+    alphabet = events log1
+    a2i s    = Set.findIndex s alphabet
+    i2a i    = Set.elemAt i alphabet
+    intLog   = map (map a2i) log1
+    
+findConcurrentPairs :: String -> IO()
+findConcurrentPairs inputLog = do
+    (logOriginal, alphabet, decode) <- fmap toIntLog $ readLog inputLog
+    let log    = dropSubtraces $ split logOriginal 
+        oracle = oracleMokhovCarmona log
+        events = Set.elems alphabet
+        cache  = Map.fromSet (uncurry oracle) $
+                 Set.fromList [ (x, y) | x <- events, y <- events, x <= y ]
+        co a b = cache Map.! (min a b, max a b)
+
+    putStrLn "Concurrent pairs:"
+    print [ (decode x, decode y) | (x:xs) <- tails events, y <- xs, x `co` y ]
+    putStrLn ""
+    putStrLn "______________"
+    putStrLn ""
+
+
+
+main :: IO ()
+main = do
+    let log1 = "a b c a c b"
```

Comment:
I think it works the same because we attempt to split traces into shorter one to avoid event repetitions. The first repetition happens exactly on the second `a`, hence the result may indeed be the same :)